hathor/consensus.py

```diff
@@ -1,5 +1,5 @@
 from itertools import chain
-from typing import TYPE_CHECKING, Iterable, List, Optional, Set
+from typing import TYPE_CHECKING, Iterable, List, Optional, Set, cast
```

Comment:
The problem I am facing now is about the metadata saved on storage.

When we load without any specific parameters, we do a fast init, where we trust that the data on the storage is correct, so we just load it. When we do a full verification, we reset the metadata of each transaction and calculate every step of the consensus for each transaction.

The problem is when I start loading my storage with a full verification and stop before it's fully initialized, so the txs that I had already loaded would have partial metadata informations because the txs not loaded could change its consensus. Then if I decide to do a fast load, I can't trust the metadata informations that are on the storage and were updated with partial information.

I am still thinking about the best solution but my first thought is to mark the metadata I reset with a full verification as 'unreliable'. It will be reliable only after a full initialization, then I can trust it. While doing a fast init, if the metadata is unreliable, I have to do the full check again.

Another option would be to have a 'temporary' storage while loading the transactions that is only updated after it's completely initialized. However this would require the double of storage to run an initialization.

Any thoughts on that? @jansegre @obiyankenobi @msbrogli 

(I created this comment here on this line to start a discussion thread, not only a single comment)

Reply:
This makes me dislike the "full verification" even more (besides being slow it can mess up your storage).

I don't think a temporary storage would be really good, it would either need not so trivial implementations for all backends or have a really expensive workaround (like when loading storage A, redirecting all writes to a memory storage and after it's loaded flushing all the writes back to storage A).

I think some sort of state, or lock would be simpler and would solve the problem. Even a lock file would be enough, this is probably going to have to be storage specific, but we lock the storage when starting a full verification, and unlock when its finished. If it's already locked when we start to load (full verification or not) we have to assume the storage is in a broken state and either a) refuse to continue, b) wipe the storage, or c) make a full verification; which are all bad and will need a long full sync to recover (unless we make periodic storage backups, which would be good for many reasons, but probably need backend specific support).

Reply:
I agree that a temporary storage is not a good solution.

Following @jansegre's idea, we already have to improve our storage to support general items (key/value). In this sense, we can set the key `runningFullVerification` to true when it starts and to false when it ends. So, if it is interrupted, we can detect in the next run. In this case, a full verification will be required and the fast init cannot be used. What do you think?

We should also create a new design to discuss how we can run a full verification without changing the metadata. I think it shouldn't be complicated, but we don't have to do it here.

Reply:
I also think we don't have to do this on this PR. Also, a way to run non-destructive verification would enable verifying only the last few txs/blocks (like the last 100 blocks, or everything in the last 1h from the last timestamp in storage). Which is much more reasonable to me than a full verification.

Reply:
I was thinking about a way to keep the unmodified storage data as clean, so we would need to calculate again only the ones modified by the full verification. But I agree that there are no good solutions for that.

With @jansegre suggestion to change the parameter to --full-verification, it's harder for someone to run this by mistake, so I guess it's fine to assume that if you stop in the middle you can't run a fast init again until you finish a full init.

Reply:
@pedroferreira1 I think it's fine either way, if you always run verifications you won't have problems or if you always run without verifications you won't have problems. You'll only have problems if you run verifications, stop in before it finishes, and run without verifications.

Maybe adding a comment on the CLI's help for `--full-verification` (like "This will break your storage if you stop midway") is enough for now.

I just don't want to delay this PR much longer.

Reply:
The changes for this are here: https://github.com/HathorNetwork/hathor-core/pull/89/commits/fad36e6a6fd02e77a89ddc825fc2320acaf063fd

@msbrogli @jansegre @obiyankenobi

Reply:
If the full node is killed for any reasons, and it can't shutdown properly, we would need to run a full verification as well. So, we need to be careful about it.

Reply:
> If the full node is killed for any reasons, and it can't shutdown properly, we would need to run a full verification as well. So, we need to be careful about it.

Then we should review our storages, it seems perfectly reasonable for our storages to be resilient against sudden crashes. RocksDB supports transactions, and we should take advantage of that so that we never leave the database in an inconsistent state.

Reply:
I think today we also have this problem with transactions (not only metatadas) and a full verification after a sudden crash wouldn't fix it. To prevent this from happen in both metadata and tx data, we would need to refactor our storage, which I think we should do, but I am just wondering if we should do here, given that this is a problem we always had.

Reply:
Definitely not on this PR, which I think has stretched too far from its original purpose already. I'm okay with some things being a little worse before we can make them a lot better, as long is it's nothing too serious.

Reply:
Issue created. https://github.com/HathorNetwork/hathor-core/issues/123